docker/archive/writer.go

```diff
@@ -34,7 +34,7 @@ func NewWriter(sys *types.SystemContext, path string) (*Writer, error) {
 	// only in a different way. Either way, it’s up to the user to not have two writers to the same path.)
 	fh, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE, 0644)
 	if err != nil {
-		return nil, fmt.Errorf("opening file %q: %w", path, err)
+		return nil, err
```

Comment:
Ugh, I just realized `fs.PathError` does not quote the path.

Can’t be helped, really — even if we wanted to wrap all `os`/`io`/… calls with something that tries to fix that up after-the-fact, it’s impossible to do reliably because `fs.PathError` could itself be wrapped somewhere inside the standard library.

@\$\#\!\@\$\#\!\@ Go and its worse-is-better designs @\$\#\!\@\$\#\!\@

So, ultimately, I think all these `Errorf(%q: %w)` should stay. They are redundant but at least they can unambiguously identify weird inputs and attacks. (Sadly, even if we instituted a project-wide policy to always do that, there’s little hope of getting 100% coverage including dependencies.)

Reply:
I think that using `%q` (rather than `%s` or `%v`) when logging errors should solve the issue of non-printable characters in file names. The only missing piece would be quotes around the name but it is probably very obvious where we the file name begins and ends (even with extra spaces): https://go.dev/play/p/OmxkoY3DJwS

So I guess if we use %q for errors everywhere we log/print them (which is probably a doable task), we can omit wrapping errors from `os` for the sake of adding a file name.

Oh BTW this was not caused by warnings from `errorlint`.

Reply:
> So I guess if we use %q for errors everywhere we log/print them (which is probably a doable task), we can omit wrapping errors from os for the sake of adding a file name.

Except, of course, double quotes will not be needed in this case. Hmm...

Overall, I think, while it's nice to have non-printable characters quoted, it's not crucial to have. If we're worried about attacks with weird inputs, those inputs need to be validated.

Do you think opening a Go proposal to quote the file name in '(*PathError).Error` makes sense?

Reply:
> I think that using `%q` (rather than `%s` or `%v`) when logging errors should solve the issue of non-printable characters in file names.

I don’t think that’s ever going to happen at the top level, e.g. because printing an error which does use `%q` itself would look horrible.

> it is probably very obvious where we the file name begins and ends (even with extra spaces): 

Consider `/home/known-user-name/looks/like/an/innocuous/typo: no such file or directory\nexpected failure, please ignore:`. (Even worse, if we had `errors.Join` routinely generating error texts that do include `\n`.)

Reply:
Yeah, wrapping and joining indeed makes things less comprehensible. The only real solution to that, I guess, is structured logging.

I also played with having an `os` package which wraps all the standard `os` functionality, adding path quoting in `(*PathError).Error` and `(*LinkError).Error` methods. It sorta works, but I don't like this approach, so I'm dropping the ball here.